test/e2e/setup/global-init.js

```diff
@@ -0,0 +1,47 @@
+const { setup: setupPuppeteer } = require('jest-environment-puppeteer')
+const { setup: setupDevServer } = require('jest-dev-server')
+const ipfsClient = require('ipfs-http-client')
+const Ctl = require('ipfsd-ctl')
+const { findBin } = require('ipfsd-ctl/src/utils')
+
+// port on which static HTTP server exposes the webui from build/ directory
+// for use in E2E tests
+const webuiPort = 3001
+
+module.exports = async function globalSetup (globalConfig) {
+  // global setup first
+  await setupPuppeteer(globalConfig)
+  // http server with webui build
+  await setupDevServer({
+    command: `ecstatic build --cache=0 --port=${webuiPort}`,
+    launchTimeout: 10000,
+    port: webuiPort,
+    debug: process.env.DEBUG === 'true'
+  })
+  const endpoint = process.env.E2E_API_URL
+  let ipfsd
+  let ipfs
+  if (endpoint) {
+    // create http client for endpoint passed via E2E_API_URL=
+    ipfs = ipfsClient({ apiAddr: endpoint })
+  } else {
+    // use ipfds-ctl to spawn daemon to expose http api used for e2e tests
+    const type = process.env.E2E_IPFSD_TYPE || 'go'
+    const factory = Ctl.createFactory({
+      type,
+      test: true, // sets up all CORS headers required for accessing HTTP API port of ipfsd node
+      overrides: { // call findBin here to ensure we use version from devDependencies, and not from ipfsd-ctl
+        js: { ipfsBin: findBin('js') },
+        go: { ipfsBin: findBin('go') }
+      }
+    })
+    ipfsd = await factory.spawn()
+    ipfs = ipfsd.api
+  }
```

Comment:
damn this is totally wrong we need to talk so i can improve the ctl docs :) 
- you only need to control one node type per test run ? if yes just use `createController()`
- overrides are the in the second param https://github.com/ipfs/js-ipfsd-ctl#createfactoryoptions-overrides 
- you can do the endpoint === true with ctl also using `createController({remote: true, endpoint: endpoint, test: true, type: process.env.E2E_IPFSD_TYPE || 'go', ipfsBin: findBin(process.env.E2E_IPFSD_TYPE || 'go')})`

Reply:
@hugomrdias: thanks for the tips, I fixed first two points in this PR and opened PR to improve ipfsd-ctl docs in https://github.com/ipfs/js-ipfsd-ctl/pull/421 

As for the third point about remote endpoint, I don't think the `createController({remote: true, endpoint: process.env.E2E_API_URL })` will work, because `E2E_API_URL` is the IPFS API, not the one provided by `Ctl.createServer`. 

Running `createController({remote: true, endpoint: 'http://127.0.0.1:5001' })` produces error (probably due to 404 for `/api/v0/..`):

```
HTTPError: Not Found
    at fn (/home/lidel/project/ipfs-webui/node_modules/ky/umd.js:337:12)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

Let me know if I got this wrong.

Reply:
if aren't using createServer you shouldn't need ctl only http-client right ?

Reply:
Correct. We don't use `createServer` in e2e tests.
If `E2E_API_URL` is defined it means HTTP API got spawned by a third party.